panel/dataclass.py

```diff
@@ -0,0 +1,357 @@
+):
+    """Syncs the fields of the model with the parameters of the Parameterized object.
+
+    Args:
+        model: The model to synchronize.
+        parameterized: The Parameterized object to synchronize.
+        names: The names of the fields/parameters to synchronize. If none are
+            specified, all public and relevant fields of the model will be synced. If a dict is
+            specified it maps from field names to parameter names.
+
+    Example:
+
+    ```python
+    import panel as pn
+    import ipyleaflet as ipyl
+    import param
+
+    pn.extension("ipywidgets")
+
+    leaflet_map = ipyl.Map()
+
+    class Map(param.Parameterized):
+        center = param.List(default=[52.204793, 360.121558])
+        zoom = param.Number(default=4, bounds=(0, 24), step=1)
+
+    parameterized = Map()
+
+    pn.dataclass.sync_with_parameterized(
+        model=leaflet_map, parameterized=parameterized
+    )
+    pn.Column(leaflet_map, parameterized.param.zoom, parameterized.param.center).servable()
+    ```
+
+    """
+    _get_utils(model).sync_with_parameterized(model, parameterized, names)
+
+def sync_with_widget(
+    model: DataClassLike,
+    widget: Widget,
+    name: str="value",
+):
+    """Syncs the named field of the model with value parameter of the widget.
+
+    Args:
+        model: The model to synchronize.
+        widget: The Panel widget to synchronize.
+        name: The name of the field to synchronize. Default is 'value'.
+
+    Example:
+
+    ```python
+    import panel as pn
+    import ipyleaflet as ipyl
+
+    pn.extension("ipywidgets")
+
+    leaflet_map = ipyl.Map()
+
+    zoom_widget = pn.widgets.FloatSlider(value=2.0, start=1.0, end=24.0, name="Zoom")
+    zoom_control_widget = pn.widgets.Checkbox(name="Show Zoom Control")
+
+    pn.dataclass.sync_with_widget(leaflet_map, zoom_widget, name="zoom")
+    pn.dataclass.sync_with_widget(leaflet_map, zoom_control_widget, name="zoom_control")
+    pn.Column(leaflet_map, zoom_widget, zoom_control_widget).servable()
+    ```
+    """
+    sync_with_parameterized(model, widget, names={name: "value"})
+
+class ModelParameterized(Parameterized):
+    """An abstract Parameterized base class for wrapping a traitlets HasTraits class or instance."""
+
+    model: DataClassLike = param.Parameter(allow_None=False, constant=True)
+    # Todo: consider renaming to names or model_names because it will be used publicly
+    _model_names: Iterable[str] | dict[str, str] = ()
+
+    def __init__(self, **params):
+        if "model" not in params and hasattr(self, "_model_class"):
+            params["model"] = self._model_class()
+
+        model = params["model"]
+        names = params.pop("names", self._model_names)
+        utils = _get_utils(model)
+        if not names:
+            names = utils.get_public_and_relevant_field_names(model)
+        names = utils.ensure_dict(names)
+        self._model_names = names
+        for parameter in names.values():
```

Comment:
I discovered that `self.param.add_parameter(parameter, param.Parameter())` adds to the class (NOT EXPECTED BY ME) and not the instance. C.f. test failure.

I need this to only add parameters to the instance. How do I do that?

Reply:
Not currently possible afaik. Why not dynamically create a class?

Reply:
Yeah I'll change this to create the class dynamically.

Reply:
Hmm, I'm exceptionally confused by this. Shouldn't `create_parameterized` be in charge of creating a subclass that creates the Parameters?